keps/sig-cli/20190920-future-of-kubectl-cp.md

```diff
@@ -0,0 +1,141 @@
+[Proposal to drop kubectl cp](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/kubernetes-sig-cli/_zUy67lK49k/aE6vncYiAgAJ).   
+
+This document aims to solidify the future of `kubectl cp` as a tool that provides basic function of copying files between local environments and pods.  Any advanced use cases
+such as those involving symlinks or modifying file permissions should be performed outside of `kubectl cp` through `kubectl exec`, addons, or shell commands.    
+
+Over the past few releases, there have been numerous security issues with `kubectl cp` that have resulted in release updates in all supported versions of kubectl.
+At the same time,any new PR that extends `kubectl cp` must undergo extra reviews to evaluate security threats that may arise [1][2].  Over the past few months,
+security fixes have required dropping edge cases and function of the command.  It is increasingly difficult to maintain a cp command that is both
+useful and secure.  There are alternative approaches that provide the same function as `kubectl cp` [3].  Using `standalone pipe + tar`
+provides transparency when copying files as well as more proven mitigations for path traversals, symlink directory escapes, tar bombs, and other exploits.
+Use of tar is also more more featureful, in that it can preserve file permissions and copy pod-to-pod.   
+
+With all of this in mind the cost and risk of maintaining the cp command should be weighed against what is considered crucial functionality in kubectl. 
+It's better to address 80% of use cases with a simple tool than trying to address the remaining 20% at the cost of risking those 80%.     
+
+[1] https://github.com/kubernetes/kubernetes/pull/78622   
+[2] https://github.com/kubernetes/kubernetes/pull/73053   
+[3] https://gist.github.com/tallclair/9217e2694b5fdf27b55d6bd1fda01b53   
+
+## Motivation
+
+- The `kubectl cp` command has been the subject of multiple security vulnerability reports.
+- Even with basic use cases of `kubectl cp`, there are long-standing issues that have not been addressed.   
+     1) container images are required to have the tar binary in order to use `kubectl cp`.    
+     2) cp command is not well maintained for Windows, if it works at all.    
+
+Given its history, `kubectl cp` should perform only basic copying of files.  Advanced features of file copying should be out of scope for `kubectl cp`.  
+
+### Goals for kubectl cp
+
+- Provide simple function to copy a single file or directory, without advanced features such as symlinks or file permission changes
+- Ensure that features added to `kubectl cp` are well understood and easy to secure.
+- Offer users example `kubectl exec`/ shell commands to address advanced options of copying files. (There are a few already in --help)
+
+### Non-Goals
+
+- Rewrite `kubectl cp` to not use tar, by modifying CRI as outlined partially [here](https://github.com/kubernetes/kubernetes/issues/58512). 
```

Comment:
what about using an ephemeral container? https://github.com/kubernetes/kubernetes/issues/58512#issuecomment-533640463
- still uses tar, but one kubernetes controls
- doesn't involve modifying CRI
- may not require dropping this feature from kubectl

Reply:
Seems like it could be good, but it will require a bunch of kubelet work, so we should weigh cost/benefit....

Reply:
added as a non-goal for this proposal, but agreed it could solve the issue of requiring tar in image

Reply:
Potentially it also makes the tar binary then relatively trusted?
What kuebelet work does this require? Didn't we already add ephemeral containers?